src/order/heyting_algebra.lean

```diff
@@ -0,0 +1,689 @@
+@[notation_class] class has_hnot (α : Type*) := (hnot : α → α)
+
+export has_himp (himp) has_sdiff (sdiff) has_hnot (hnot)
+
+infixr ` ⇨ `:60 := himp
+prefix `￢`:72 := hnot
+
+instance [has_himp α] [has_himp β] : has_himp (α × β) := ⟨λ a b, (a.1 ⇨ b.1, a.2 ⇨ b.2)⟩
+instance [has_hnot α] [has_hnot β] : has_hnot (α × β) := ⟨λ a, (￢a.1, ￢a.2)⟩
+instance [has_sdiff α] [has_sdiff β] : has_sdiff (α × β) := ⟨λ a b, (a.1 \ b.1, a.2 \ b.2)⟩
+instance [has_compl α] [has_compl β] : has_compl (α × β) := ⟨λ a, (a.1ᶜ, a.2ᶜ)⟩
+
+@[simp] lemma fst_himp [has_himp α] [has_himp β] (a b : α × β) : (a ⇨ b).1 = a.1 ⇨ b.1 := rfl
+@[simp] lemma snd_himp [has_himp α] [has_himp β] (a b : α × β) : (a ⇨ b).2 = a.2 ⇨ b.2 := rfl
+@[simp] lemma fst_hnot [has_hnot α] [has_hnot β] (a : α × β) : (￢a).1 = ￢a.1 := rfl
+@[simp] lemma snd_hnot [has_hnot α] [has_hnot β] (a : α × β) : (￢a).2 = ￢a.2 := rfl
+@[simp] lemma fst_sdiff [has_sdiff α] [has_sdiff β] (a b : α × β) : (a \ b).1 = a.1 \ b.1 := rfl
+@[simp] lemma snd_sdiff [has_sdiff α] [has_sdiff β] (a b : α × β) : (a \ b).2 = a.2 \ b.2 := rfl
+@[simp] lemma fst_compl [has_compl α] [has_compl β] (a : α × β) : aᶜ.1 = a.1ᶜ := rfl
+@[simp] lemma snd_compl [has_compl α] [has_compl β] (a : α × β) : aᶜ.2 = a.2ᶜ := rfl
+
+namespace pi
+variables {π : ι → Type*}
+
+instance [Π i, has_himp (π i)] : has_himp (Π i, π i) := ⟨λ a b i, a i ⇨ b i⟩
+instance [Π i, has_hnot (π i)] : has_hnot (Π i, π i) := ⟨λ a i, ￢a i⟩
+
+lemma himp_def [Π i, has_himp (π i)] (a b : Π i, π i) : (a ⇨ b) = λ i, a i ⇨ b i := rfl
+lemma hnot_def [Π i, has_hnot (π i)] (a : Π i, π i) : ￢a = λ i, ￢a i := rfl
+
+@[simp] lemma himp_apply [Π i, has_himp (π i)] (a b : Π i, π i) (i : ι) : (a ⇨ b) i = a i ⇨ b i :=
+rfl
+@[simp] lemma hnot_apply [Π i, has_hnot (π i)] (a : Π i, π i) (i : ι) : (￢a) i = ￢a i := rfl
+
+end pi
+
+/-- A generalized Heyting algebra is a lattice with an additional binary operation `⇨` called
+Heyting implication such that `a ⇨` is right adjoint to `a ⊓`.
+
+ This generalizes `heyting_algebra` by not requiring a bottom element. -/
+class generalized_heyting_algebra (α : Type*) extends lattice α, has_top α, has_himp α :=
+(le_top : ∀ a : α, a ≤ ⊤)
+(le_himp_iff (a b c : α) : a ≤ b ⇨ c ↔ a ⊓ b ≤ c)
+
+/-- A generalized co-Heyting algebra is a lattice with an additional binary difference operation `\`
+such that `\ a` is right adjoint to `⊔ a`.
+
+This generalizes `coheyting_algebra` by not requiring a top element. -/
+class generalized_coheyting_algebra (α : Type*) extends lattice α, has_bot α, has_sdiff α :=
+(bot_le : ∀ a : α, ⊥ ≤ a)
+(sdiff_le_iff (a b c : α) : a \ b ≤ c ↔ a ≤ b ⊔ c)
+
+/-- A Heyting algebra is a bounded lattice with an additional binary operation `⇨` called Heyting
+implication such that `a ⇨` is right adjoint to `a ⊓`. -/
+class heyting_algebra (α : Type*) extends generalized_heyting_algebra α, has_bot α, has_compl α :=
+(bot_le : ∀ a : α, ⊥ ≤ a)
+(himp_bot (a : α) : a ⇨ ⊥ = aᶜ)
```

Comment:
The following
```suggestion
class heyting_algebra (α : Type*) extends generalized_heyting_algebra α, has_bot α, has_compl α :=
(compl := λ a : α, a ⇨ ⊥)
(bot_le : ∀ a : α, ⊥ ≤ a)
(himp_bot (a : α) : a ⇨ ⊥ = aᶜ . obviously)
```
gives the error
```
failed to add declaration 'heyting_algebra.compl._default' to environment, type has metavariables
remark: set 'formatter.hide_full_terms' to false to see the complete term
  …
```
Does anyone understand why? The trace gives
```
failed to add declaration 'heyting_algebra.compl._default' to environment, type has metavariables
  Π {α : Type u_4} (sup : α → α → α) (le lt : α → α → Prop) (le_refl : ∀ (a : α), a ≤ a)
  (le_trans : ∀ (a b c : α), a ≤ b → b ≤ c → a ≤ c)
  (lt_iff_le_not_le :
    auto_param (∀ (a b : α), a < b ↔ a ≤ b ∧ ¬b ≤ a) (name.mk_string "order_laws_tac" name.anonymous))
  (le_antisymm : ∀ (a b : α), a ≤ b → b ≤ a → a = b) (le_sup_left : ∀ (a b : α), a ≤ a ⊔ b)
  (le_sup_right : ∀ (a b : α), b ≤ a ⊔ b) (sup_le : ∀ (a b c : α), a ≤ c → b ≤ c → a ⊔ b ≤ c)
  (inf : α → α → α) (inf_le_left : ∀ (a b : α), a ⊓ b ≤ a) (inf_le_right : ∀ (a b : α), a ⊓ b ≤ b)
  (le_inf : ∀ (a b c : α), a ≤ b → a ≤ c → a ≤ b ⊓ c) (top : α) (himp : α → α → α),
    (∀ (a : α), a ≤ ⊤) → (∀ (a b c : α), a ≤ b ⇨ c ↔ a ⊓ b ≤ c) → α → α → α
```
which looks like a fine term to me.

Reply:
It's kind of suspicious that it depends on so many things beyond `himp`, but I have no idea if there's a good way to get rid of all of those, or if that would help.

Reply:
Isn't it simply because of the way structures are encoded? This seems to be all fields that are declared before `compl`.

Reply:
I think the problem really comes from the extra parameters. For example, if we use the names of the structure fields instead of notation, we don't get the error (and the type becomes much smaller):
```suggestion
class heyting_algebra (α : Type*) extends generalized_heyting_algebra α, has_bot α, has_compl α :=
(compl := λ a : α, himp a bot)
(bot_le : ∀ a : α, ⊥ ≤ a)
(himp_bot (a : α) : a ⇨ ⊥ = aᶜ)
```

Reply:
[Zulip thread on this error](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/.22type.20has.20metavariables.22.20error.20in.20structure.20default.20field)